File: architecture-skills/design-patterns/examples/state/state.swift
// State pattern in Swift.
//
// An enum with mutating methods is the idiomatic Swift state machine:
// exhaustive switch means the compiler flags unhandled (state, event)
// pairs, and associated values can carry per-state data. Use the
// protocol/class form only when states are open-ended.

enum MachineState {
    case idle
    case hasCoin
    case soldOut
}

struct VendingMachine {
    private(set) var state: MachineState
    private(set) var stock: Int

    init(stock: Int) {
        self.stock = stock
        self.state = stock > 0 ? .idle : .soldOut
    }

    mutating func insertCoin() {
        switch state {
        case .idle:
            print("coin accepted")
            state = .hasCoin
        case .hasCoin:
            print("coin already inserted, returning it")
        case .soldOut:
            print("sold out, returning coin")
        }
    }

    mutating func selectItem() {
        switch state {
        case .idle:
            print("insert a coin first")
        case .hasCoin:
            stock -= 1
            print("item dispensed (\(stock) left)")
            state = stock == 0 ? .soldOut : .idle
        case .soldOut:
            print("sold out")
        }
    }
}

var machine = VendingMachine(stock: 2)
machine.selectItem() // insert a coin first
machine.insertCoin()
machine.insertCoin() // coin already inserted
machine.selectItem() // item dispensed (1 left)
machine.insertCoin()
machine.selectItem() // item dispensed (0 left)
machine.insertCoin() // sold out, returning coin

File: architecture-skills/design-patterns/examples/state/state.rs
// State pattern in Rust.
//
// An enum + match is the idiomatic Rust state machine: the compiler
// checks that every (state, event) pair is handled, and states can
// carry data in their variants. The GoF trait-object form
// (Box<dyn State> with a trait per behavior) exists too, but is only
// worth it when states are open-ended or defined across crates.

#[derive(Debug, Clone, Copy, PartialEq)]
enum MachineState {
    Idle,
    HasCoin,
    SoldOut,
}

struct VendingMachine {
    state: MachineState,
    stock: u32,
}

impl VendingMachine {
    fn new(stock: u32) -> Self {
        Self {
            state: if stock > 0 {
                MachineState::Idle
            } else {
                MachineState::SoldOut
            },
            stock,
        }
    }

    fn insert_coin(&mut self) {
        self.state = match self.state {
            MachineState::Idle => {
                println!("coin accepted");
                MachineState::HasCoin
            }
            MachineState::HasCoin => {
                println!("coin already inserted, returning it");
                MachineState::HasCoin
            }
            MachineState::SoldOut => {
                println!("sold out, returning coin");
                MachineState::SoldOut
            }
        };
    }

    fn select_item(&mut self) {
        self.state = match self.state {
            MachineState::Idle => {
                println!("insert a coin first");
                MachineState::Idle
            }
            MachineState::HasCoin => {
                self.stock -= 1;
                println!("item dispensed ({} left)", self.stock);
                if self.stock == 0 {
                    MachineState::SoldOut
                } else {
                    MachineState::Idle
                }
            }
            MachineState::SoldOut => {
                println!("sold out");
                MachineState::SoldOut
            }
        };
    }
}

fn main() {
    let mut machine = VendingMachine::new(2);
    machine.select_item(); // insert a coin first
    machine.insert_coin();
    machine.insert_coin(); // coin already inserted
    machine.select_item(); // item dispensed (1 left)
    machine.insert_coin();
    machine.select_item(); // item dispensed (0 left)
    machine.insert_coin(); // sold out, returning coin
}
